### ISSUE TITLE
Qwen2.5MoE load error

### ISSUE THREAD

**Issue by sempervictus** [EricLBuehler/candle-vllm#292]
Think the loader might need a bit of the "safe defaults" treatment when loading `DavidAU/Qwen2.5-8x7B-Vee-Eight-Coder-Instruct-53B-128k-ctx`:
```rust
candle-vllm-swe  | 2025-09-02T16:35:46.575733Z  INFO candle_vllm::openai::pipelines::pipeline: Model Config { architectures: Some(["Qwen2MoeForCausalLM"]), hidden_size: 3584, head_dim: Some(128), intermediate_size: 18944, vocab_size: 152064, num_hidden_layers: 28, num_attention_heads: 28, num_key_value_heads: Some(4), rms_norm_eps: 1e-6, rope_theta: 1000000.0, rope_local_base_freq: None, bos_token_id: Some(TokenID(Left(Some(151643)))), eos_token_id: TokenID(Left(Some(151645))), max_seq_len: 131072, original_max_position_embeddings: None, sliding_window: None, sliding_window_pattern: None, hidden_act: Some(Silu), hidden_activation: None, tie_word_embeddings: false, rope_scaling: Some({"type": String("yarn"), "factor": Single(4.0), "original_max_position_embeddings": Single(32768.0)}), max_position_embeddings: Some(131072), attention_bias: Some(true), partial_rotary_factor: None, qk_layernorm: false, use_qkv_bias: None, custom_stop_tokens: None, attn_logit_softcapping: None, final_logit_softcapping: None, moe_config: Some(QwenMoE(QwenMoEConfig { moe_intermediate_size: 18944, shared_expert_intermediate_size: Some(18944), num_experts: Some(8), mlp_only_layers: Some([]), decoder_sparse_step: Some(1), norm_topk_prob: true, num_experts_per_tok: 2 })), quantization_config: None, quant: Some("q8_0") }
candle-vllm-swe  | 2025-09-02T16:35:46.575761Z  INFO candle_vllm::openai::pipelines::pipeline: Loading Qwen2MoeForCausalLM model.
candle-vllm-swe  | 2025-09-02T16:35:46.575830Z  INFO candle_vllm::openai::communicator: build command channel for the main process!
candle-vllm-swe  | 2025-09-02T16:35:46.876097Z  WARN candle_vllm::openai::pipelines::pipeline: create nccl comm channel rank 0, shards 4, id Id { id: ncclUniqueId { internal: [-9, 42, -118, -98, 108, -92, -119, -75, 2, 0, -24, 89, -84, 26, 0, 5, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0] } }
...
candle-vllm-swe  | 2025-09-02T16:35:47.575842Z  INFO candle_vllm::openai::communicator: one daemon process connected to the command channel!
Loading model in 4 ranks!
candle-vllm-swe  | [00:00:00] ------------------------------------------------------------    0/28   On Rank 0 Device
candle-vllm-swe  | [00:00:00] ------------------------------------------------------------    0/28   On Rank 1 Device
candle-vllm-swe  | [00:00:00] ------------------------------------------------------------    0/28   On Rank 2 Device
candle-vllm-swe  | [00:00:00] ------------------------------------------------------------    0/28   On Rank 3 Device                                                              
candle-vllm-swe  | thread 'main' (1) panicked at /candle-vllm/src/openai/pipelines/pipeline.rs:625:34:
candle-vllm-swe  | called `Result::unwrap()` on an `Err` value: cannot find tensor model.layers.0.self_attn.q_norm.weight
candle-vllm-swe  |    0: candle_core::error::Error::bt
candle-vllm-swe  |    1: candle_core::safetensors::MmapedSafetensors::get
candle-vllm-swe  |    2: candle_core::safetensors::MmapedSafetensors::load
candle-vllm-swe  |    3: <candle_core::safetensors::MmapedSafetensors as candle_nn::var_builder::SimpleBackend>::get
candle-vllm-swe  |    4: <candle_nn::var_builder::ShardedSafeTensors as candle_nn::var_builder::Backend>::get
candle-vllm-swe  |    5: candle_nn::var_builder::VarBuilderArgs<B>::get_with_hints_dtype
candle-vllm-swe  |    6: candle_vllm::openai::models::qwen3_moe::Qwen3MoE::new
candle-vllm-swe  |    7: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &F>::call_mut
candle-vllm-swe  |    8: rayon::iter::plumbing::Folder::consume_iter
candle-vllm-swe  |    9: rayon::iter::plumbing::bridge_producer_consumer::helper
candle-vllm-swe  |   10: rayon::iter::collect::collect_with_consumer
candle-vllm-swe  |   11: candle_vllm::openai::pipelines::pipeline::DefaultLoader::load_model::{{closure}}
candle-vllm-swe  |   12: candle_vllm::main::{{closure}}
candle-vllm-swe  |   13: tokio::runtime::park::CachedParkThread::block_on
candle-vllm-swe  |   14: candle_vllm::main
candle-vllm-swe  |   15: std::sys::backtrace::__rust_begin_short_backtrace
candle-vllm-swe  |   16: std::rt::lang_start::{{closure}}
candle-vllm-swe  |   17: std::rt::lang_start_internal
candle-vllm-swe  |   18: main
candle-vllm-swe  |   19: <unknown>
candle-vllm-swe  |   20: __libc_start_main
candle-vllm-swe  |   21: _start
```
ping @guoqingbao re #291

**Comment by guoqingbao** [comment]
> [@guoqingbao](https://github.com/guoqingbao) the qwen2moe i tried loading above crashed on-boot with a different error - looks like we have something backwards here:
> 
> candle-vllm-swe  | called `Result::unwrap()` on an `Err` value: shape mismatch for model.layers.0.mlp.shared_expert_gate.weight, expected: [3584], got: [1, 3584]
> candle-vllm-swe  |    0: candle_core::error::Error::bt
> candle-vllm-swe  |    1: <candle_core::safetensors::MmapedSafetensors as candle_nn::var_builder::SimpleBackend>::get
> candle-vllm-swe  |    2: <candle_nn::var_builder::ShardedSafeTensors as candle_nn::var_builder::Backend>::get
> candle-vllm-swe  |    3: candle_nn::var_builder::VarBuilderArgs<B>::get_with_hints_dtype
> candle-vllm-swe  |    4: candle_vllm::openai::models::qwen3_moe::Qwen3MoE::new
> candle-vllm-swe  |    5: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &F>::call_mut
> candle-vllm-swe  |    6: rayon::iter::plumbing::Folder::consume_iter
> candle-vllm-swe  |    7: rayon::iter::plumbing::bridge_producer_consumer::helper
> candle-vllm-swe  |    8: rayon::iter::collect::collect_with_consumer
> candle-vllm-swe  |    9: candle_vllm::openai::pipelines::pipeline::DefaultLoader::load_model::{{closure}}
> candle-vllm-swe  |   10: candle_vllm::main::{{closure}}
> ..
> qwen3moe seems to be loading for
> 
> candle-vllm-swe  | 2025-09-03T04:23:21.627133Z  INFO candle_vllm::openai::pipelines::pipeline: Model Config { architectures: Some(["Qwen3MoeForCausalLM"]), hidden_size: 2048, head_dim: Some(128), intermediate_size: 6144, vocab_size: 151936, num_hidden_layers: 84, num_attention_heads: 32, num_key_value_heads: Some(4), rms_norm_eps: 1e-6, rope_theta: 10000000.0, rope_local_base_freq: None, bos_token_id: Some(TokenID(Left(Some(151643)))), eos_token_id: TokenID(Left(Some(151645))), max_seq_len: 262144, original_max_position_embeddings: None, sliding_window: None, sliding_window_pattern: None, hidden_act: Some(Silu), hidden_activation: None, tie_word_embeddings: false, rope_scaling: None, max_position_embeddings: Some(262144), attention_bias: Some(false), partial_rotary_factor: None, qk_layernorm: false, use_qkv_bias: None, custom_stop_tokens: None, attn_logit_softcapping: None, final_logit_softcapping: None, moe_config: Some(QwenMoE(QwenMoEConfig { moe_intermediate_size: 768, shared_expert_intermediate_size: None, num_experts: Some(128), mlp_only_layers: Some([]), decoder_sparse_step: Some(1), norm_topk_prob: true, num_experts_per_tok: 8 })), quantization_config: None, quant: Some("q8_0") }
> (will see what happens after it completes quant)

May fault, the quantized qwen2 moe has different shape with unquantized one in regarding to shared expert gates. Will fix it now.

**Comment by guoqingbao** [comment]
> Thank you sir, will report back on the q3moe as it finishes

Are you able to test this #294 for unquantized qwen2 moe models.

**Comment by guoqingbao** [comment]
> Just tried :-) - compile error

You are quick, I just fixed the compilation error. You may retry.

**Comment by sempervictus** [comment]
Thank you very much @guoqingbao - `DavidAU/Qwen2.5-8x7B-Vee-Eight-Coder-Instruct-53B-128k-ctx` seems to be somehow internally broken:
```rust
candle-vllm-swe  | 2025-09-03T12:51:58.406877Z  WARN candle_vllm::openai::pipelines::llm_engine: Sending 1 tasks to 3 subprocesses
candle-vllm-swe  | 
candle-vllm-swe  | thread 'tokio-runtime-worker' (838) panicked at src/openai/logits_processor.rs:151:51:
candle-vllm-swe  | called `Result::unwrap()` on an `Err` value: A weight is negative, too large or not a valid number
candle-vllm-swe  |    0: candle_core::error::Error::bt
candle-vllm-swe  |    1: candle_vllm::openai::logits_processor::LogitsProcessor::sample_multinomial
candle-vllm-swe  |    2: core::ops::function::impls::<impl core::ops::function::FnMut<A> for &F>::call_mut
candle-vllm-swe  |    3: rayon::iter::plumbing::Folder::consume_iter
candle-vllm-swe  |    4: rayon::iter::plumbing::bridge_producer_consumer::helper
candle-vllm-swe  |    5: rayon::iter::collect::collect_with_consumer
candle-vllm-swe  |    6: rayon::iter::extend::<impl rayon::iter::ParallelExtend<T> for alloc::vec::Vec<T>>::par_extend
candle-vllm-swe  |    7: rayon::iter::from_par_iter::collect_extended
candle-vllm-swe  |    8: candle_vllm::openai::logits_processor::LogitsProcessor::sample
candle-vllm-swe  |    9: candle_vllm::openai::pipelines::pipeline::DefaultPipeline::sample
candle-vllm-swe  |   10: candle_vllm::openai::pipelines::llm_engine::LLMEngine::generate_once
candle-vllm-swe  |   11: rayon::iter::plumbing::Folder::consume_iter
candle-vllm-swe  |   12: rayon::iter::plumbing::bridge_producer_consumer::helper
candle-vllm-swe  |   13: rayon::iter::collect::collect_with_consumer
candle-vllm-swe  |   14: tokio::runtime::handle::Handle::block_on
candle-vllm-swe  |   15: tokio::runtime::task::raw::poll
candle-vllm-swe  |   16: std::sys::backtrace::__rust_begin_short_backtrace
candle-vllm-swe  |   17: core::ops::function::FnOnce::call_once{{vtable.shim}}
candle-vllm-swe  |   18: std::sys::pal::unix::thread::Thread::new::thread_start
```
but `DavidAU/Qwen2.5-2X11B-CODER-Dueling-Wolverines-28B` and `Qwen/Qwen2-57B-A14B-Instruct` work at ~40 and ~50 T/S respectively _on V100s at `q8_0`_

**Comment by guoqingbao** [comment]
> but `DavidAU/Qwen2.5-2X11B-CODER-Dueling-Wolverines-28B` and `Qwen/Qwen2-57B-A14B-Instruct` work at ~40 and ~50 T/S respectively _on V100s at `q8_0`_

The decoding speed looks nice, how about the accuracy? the output makes sense?
